src/transformers/models/llama4/modeling_llama4.py

```diff
@@ -779,7 +780,7 @@ def _update_causal_mask(
                 attention_mask = make_flex_block_causal_mask(
                     attention_mask,
                     query_length=sequence_length,
-                    key_length=past_key_values.get_max_cache_shape(),
+                    key_length=past_key_values.get_max_cache_shape() if past_key_values else sequence_length,
```

Comment:
```suggestion
                    key_length=past_key_values.get_max_cache_shape(),
```
there was#37327 that should have already fixed this part!

Reply:
Without the guard, there still could be issues, e.g. when passing `use_cache=False` (as past_kv gets None).

I don't think sequence length is correct here tho, especially when decoding (even if it's possibly broken atm, we don't need to make it even more broken :D). Wouldn't it make more sense to use `target_length` here? Then we could also avoid the guard. Needs to fix this probably tho: 
https://github.com/huggingface/transformers/blob/debfe904c94d1104e94c71d2276a0d522f0f5081/src/transformers/models/llama4/modeling_llama4.py#L769-L770

`target_length = past_key_values.get_max_cache_shape()`

Reply:
ah actually for shorter context this might be better

Reply:
because hybrid cache would be smaller than max chunk

Reply:
we can't call `.get_max_cache_shape()`. in past_key values if it's None though, perhaps this instead?
```
key_length=past_key_values.get_max_cache_shape() if past_key_values else None,
```

Reply:
If you look into the code, that's exactly what `target_length` is calculating beforehand (minus the missing self call "()"). That's why I suggested using `target_length` directly (with the fix on also really calling the function). See https://github.com/huggingface/transformers/blob/08f36771b33d246986d9338a729fc4ef258b999d/src/transformers/models/llama4/modeling_llama4.py#L769-L772

If we pass `None` to the flex attn mask creation, we also get what target length would compute, no?

Not exactly sure what you mean by "ah actually for shorter context this might be better" - ie that we use the max cache shape?

Reply:
thanks, updated!

Reply:
I don't think using `target length` is correct or we also need to pass that to query_length too. Using that results in this which doesn't happen when using `sequence_length`:

```
[rank1]:   File "/root/miniconda3/envs/py3.11/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 1157, in flex_attention
[rank1]:     def flex_attention(                  
[rank1]:   File "/root/miniconda3/envs/py3.11/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 1285, in torch_dynamo_resume_in_flex_attention_at_1285
[rank1]:     if query.size(-2) > block_mask_q_len or key.size(-2) > block_mask_kv_len:
[rank1]:   File "/root/miniconda3/envs/py3.11/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 1293, in torch_dynamo_resume_in_flex_attention_at_1292
[rank1]:     raise ValueError(                                                                                                
[rank1]: ValueError: block_mask was created for block_mask.shape=(1, 1, 4096, tensor(8192, device='cuda:1')) but got q_len=4096 and kv_len=4096. As the block mask was created for a larger length than you're using it for, you can either 1. create a new block mask with the correct length, or 2. 'adjust' the existing block mask to the correct length by calling block_mask._adjust(q_len, kv_len). This essentially 'crops' the block mask to the upper left corner, which does not work for all mask_mods!
```

It seems both query_length and key_length have to be consistent and we should either:
1. pass target_length to both
2. pass sequence_length to both
3. use block_mask._adjust (???)

Reply:
using `target_length` isn't quite correct either resulting in:
```
[rank7]:   File "/root/miniconda3/envs/py3.11/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 1157, in flex_attention
[rank7]:     def flex_attention(                  
[rank7]:   File "/root/miniconda3/envs/py3.11/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 1285, in torch_dynamo_resume_in_flex_attention_at_1285
[rank7]:     if query.size(-2) > block_mask_q_len or key.size(-2) > block_mask_kv_len:               
[rank7]:   File "/root/miniconda3/envs/py3.11/lib/python3.11/site-packages/torch/nn/attention/flex_attention.py", line 1293, in torch_dynamo_resume_in_flex_attention_at_1292
[rank7]:     raise ValueError(        
[rank7]: ValueError: block_mask was created for block_mask.shape=(1, 1, 4096, tensor(8192, device='cuda:7')) but got q_len=4096 and kv_len=4096. As the block mask was created for a larger length than you're using it for, you can either 1. create a new b
lock mask with the correct length, or 2. 'adjust' the existing block mask to the correct length by calling block_mask._adjust(q_len, kv_len). This essentially 'crops' the block mask to the upper left corner, which does not work for all mask_mods!
```

Reply:
Is this caused within the flex attention forward? ie https://github.com/huggingface/transformers/blob/08f36771b33d246986d9338a729fc4ef258b999d/src/transformers/integrations/flex_attention.py#L185

Is your usecase training? Then it makes sense that `kv_len == q_len`. To me, it seems that the block mask must have the exact correct shape for its input q and kv (which I wasn't completely aware of). This means we either crop it (within the forward) or pass the correct expected length. I fear inference will need a bit more work 😄

Reply:
Yeah flex attention forward. Its kinda expected because target length is not aware of the cache max allocated length basically

Reply:
yeah, my use case is training. We should probably put this PR on hold for the minute until I can do some more verification of the correct fix

Reply:
Hmm, what default value would you suggest in case `past_key_values == None` then?

Reply:
For training its just gonna be sequence length